### DISCUSSION TITLE
why Lightweight DELETEs is slower than alter delete

### DISCUSSION THREAD

**Discussion by XingYuanHuang** [ClickHouse/ClickHouse#49767] [Q&A]
I have the following table:

CREATE TABLE students
(
    `class_id` String,
    `student_id` string,
    `is_matched` Int8
)
ENGINE = MergeTree
PARTITION BY (class_id)
ORDER BY (class_id,student_id)
SETTINGS index_granularity = 8192;

This table contains 10 million rows of random data. 
Now, I'm trying to use lightweight deletion to replace the original ALTER DELETE.

However, when I execute two types of deletions concurrently at random partitions (based on class_id), using golang and sqlx, I get the following results:

All 10 'alter' operations completed in 163.9897ms
All 10 'delete' operations completed in 2.4888245s

I found that the lightweight deletion takes longer than the ALTER operation. Is this because ClickHouse needs to track which rows have been deleted for lightweight deletion?

If the overhead of lightweight deletion itself already exceeds that of the ALTER operation, I am not clear about what scenarios this feature should be used in. 

Can anyone provide some insights?

**Comment by Alekky09** [comment]
Did you use `mutations_sync` setting in your `ALTER DELETE` query? Otherwise it's async and it doesn't wait for the actual DELETE to happen.

**Reply by XingYuanHuang** [reply]
I didn't employ the mutations_sync setting, so the ALTER DELETE operation is indeed asynchronous. 
My confusion stems from the observation that the DELETE FROM statement, which should also be asynchronous, seems to take much longer than ALTER DELETE. Could this be due to the need to flag deletions in lightweight deletions? However, this overhead seems quite substantial.

**Reply by Alekky09** [reply]
I suggest you use the mutations_sync=1 for both queries and compare the results. Docs for both of the functions say that they're async in some way, but are unclear on what exactly is the difference in their sync part.

I also see that your table has PARTITION BY class_id and you're trying to delete based on that column, so you can also try DROP PARTITION, it should be more efficient since it doesn't do read on data.

**Comment by leticiawebb** [comment]
There is a blog post [Handling Updates and Deletes in ClickHouse](https://clickhouse.com/blog/handling-updates-and-deletes-in-clickhouse) that explains more about the difference between the two. The short version is:
- ALTER... DELETE begins a mutation with the sole purpose of deleting the record
- Lightweight delete waits for the next mutation to delete the record; it reduces overhead by combining the delete mutation with other operations that need to be performed in the database
